File: server/cd_search.R
# server/cd_search.R
# ============================================
# 最终版：NCBI Batch CD-Search Web API（全程验证通过）
#   submit(tdata 触发文本) -> #cdsid -> poll(#status) -> fetch -> 解析(洗 Query ID)
#   批大小 950；当前工作目录下 cd_cache/<文件名>/ 存 chunk rds + progress.txt + 自动导出 TSV
#   断点续跑；跨平台 curl；保留一蛋白多行命中
# ============================================

message("[DEBUG] cd_search.R loading... (Final + auto TSV export)")

# ---------- 配置 ----------
CD_USE_PROXY <- FALSE                          # ← 测试直连设 FALSE；走代理设 TRUE
CD_PROXY     <- "socks5h://127.0.0.1:10808"
CD_PROXY_ARG <- if (CD_USE_PROXY) paste("-x", CD_PROXY) else ""
CD_ENDPOINT <- "https://www.ncbi.nlm.nih.gov/Structure/bwrpsb/bwrpsb.cgi"
CD_CHUNK    <- 950                           # 1000 会被 NCBI 判超限，必须 <1000
CD_CURL     <- if (.Platform$OS.type == "windows") "curl.exe" else "curl"

# 自动：当前工作目录下建 cd_cache/ 总缓存目录
CD_CACHE_DIR <- normalizePath(file.path(getwd(), "cd_cache"), mustWork = FALSE)
dir.create(CD_CACHE_DIR, showWarnings = FALSE, recursive = TRUE)
message("[DEBUG] cd_search: cache root = ", CD_CACHE_DIR)

# 用上传文件名作为任务子目录名（去扩展名 + 清理非法字符）
make_task_id <- function(fname) {
  base <- tools::file_path_sans_ext(basename(fname))   # Ptenuiflora_genome_V1.prot.fasta -> Ptenuiflora_genome_V1.prot
  gsub('[\\\\/:*?"<>|]', "_", base)
}

# ---------- 读取 FASTA ----------
cd_fasta <- reactive({
  req(input$cd_search_fasta)
  file_path <- input$cd_search_fasta$datapath
  message("[DEBUG] cd_search: reading FASTA from ", file_path)
  
  if (!requireNamespace("seqinr", quietly = TRUE)) {
    showNotification("Package 'seqinr' is required.", type = "error")
    return(NULL)
  }
  tryCatch({
    seqs <- seqinr::read.fasta(file_path, seqtype = "AA", as.string = TRUE)
    ids  <- names(seqs)
    sequences <- unlist(seqs)
    message("[DEBUG] cd_search: loaded ", length(ids), " sequences")
    data.frame(ID = ids, Sequence = sequences, stringsAsFactors = FALSE)
  }, error = function(e) {
    message("[ERROR] cd_search: ", conditionMessage(e))
    showNotification(paste("Failed to read FASTA:", conditionMessage(e)), type = "error")
    NULL
  })
})

# ---------- 1) 提交，返回 cdsid（tdata 触发文本；超限硬失败，临时错误退避重试） ----------
cd_submit <- function(temp_fasta, db = "cdd", smode = "auto", retries = 3) {
  for (attempt in seq_len(retries)) {
    cmd <- sprintf(
      paste0('%s %s -s -X POST "%s" ',
             '--data-urlencode "queries@%s" ',
             '--data "db=%s" --data "smode=%s" ',
             '--data "tdata=hits" --data "dmode=rep" ',
             '--data "cddefl=false" --data "qdefl=false" --data "clonly=false" ',
             '--data "useid1=true" --max-time 120'),
      CD_CURL, CD_PROXY_ARG, CD_ENDPOINT, temp_fasta, db, smode
    )
    out <- tryCatch(system(cmd, intern = TRUE, ignore.stderr = FALSE),
                    error = function(e) character(0))
    txt <- paste(out, collapse = "\n")
    
    m <- regmatches(txt, regexpr("QM3-qcdsearch-[A-Za-z0-9-]+", txt))
    if (length(m) > 0) return(list(success = TRUE, cdsid = m[1]))
    
    if (grepl("too many|1000 or less", txt, ignore.case = TRUE)) {
      return(list(success = FALSE,
                  error = "数量超限（NCBI 拒绝），请把 CD_CHUNK 降到更小（如 900）"))
    }
    transient <- nchar(txt) == 0 ||
      grepl("temporarily|try again|service unavailable|busy", txt, ignore.case = TRUE)
    if (transient && attempt < retries) { Sys.sleep(30 * attempt); next }
    return(list(success = FALSE, error = paste("提交被拒，预览:", substr(txt, 1, 300))))
  }
  list(success = FALSE, error = "重试多次仍失败")
}

# ---------- 2) 轮询 + 取结果（用 #status 判断） ----------
cd_poll_fetch <- function(cdsid, max_wait = 900, interval = 20) {
  elapsed <- 0
  repeat {
    cmd <- sprintf(
      paste0('%s %s -s -X POST "%s" --data "cdsid=%s" --data "tdata=hits" ',
             '--data "cddefl=true" --data "qdefl=true" --data "dmode=rep" ',
             '--data "clonly=false" --max-time 120'),
      CD_CURL, CD_PROXY_ARG, CD_ENDPOINT, cdsid
    )
    out <- tryCatch(system(cmd, intern = TRUE, ignore.stderr = FALSE),
                    error = function(e) character(0))
    if (length(out) == 0) {
      Sys.sleep(interval); elapsed <- elapsed + interval
      if (elapsed >= max_wait) return(list(success = FALSE, error = "轮询无响应超时"))
      next
    }
    
    status_line <- grep("^#status", out, value = TRUE)
    status <- if (length(status_line)) {
      as.integer(sub(".*#status[^0-9]*([0-9]+).*", "\\1", status_line[1]))
    } else NA_integer_
    
    if (!is.na(status) && status == 0) {
      return(list(success = TRUE, content = paste(out, collapse = "\n")))
    }
    if (!is.na(status) && status == 3) {
      message("[DEBUG] running... (", elapsed, "s) cdsid=", cdsid)
      Sys.sleep(interval); elapsed <- elapsed + interval
      if (elapsed >= max_wait) return(list(success = FALSE, error = "等待结果超时"))
      next
    }
    if (is.na(status)) {
      Sys.sleep(interval); elapsed <- elapsed + interval
      if (elapsed >= max_wait) return(list(success = FALSE, error = "未读到 #status"))
      next
    }
    return(list(success = FALSE, error = paste("NCBI status =", status)))
  }
}

# ---------- TSV 文本 -> data.frame（去 # 注释行；洗 Query 列 ID；空命中返回 0 行） ----------
cd_parse_tsv <- function(content) {
  lines <- strsplit(content, "\n")[[1]]
  data_lines <- lines[!grepl("^#", lines) & nchar(trimws(lines)) > 0]
  if (length(data_lines) < 1) return(NULL)
  df <- tryCatch(
    read.csv(text = paste(data_lines, collapse = "\n"),
             sep = "\t", header = TRUE, stringsAsFactors = FALSE, check.names = FALSE),
    error = function(e) NULL
  )
  if (is.null(df) || ncol(df) < 2) return(NULL)
  qcol <- names(df)[1]
  df[[qcol]] <- trimws(sub("^Q#\\d+\\s*-\\s*>?", "", df[[qcol]]))   # Q#1 - >Pt_Chr01 -> Pt_Chr01
  df
}

# ============================================
# 主按钮：自动分批 + 断点续跑 + 进度文件 + 自动导出 TSV
# ============================================
observeEvent(input$run_cd_search, {
  message("[DEBUG] cd_search: button clicked")
  fasta_df <- cd_fasta()
  if (is.null(fasta_df) || nrow(fasta_df) == 0) {
    showNotification("Please upload a FASTA file first.", type = "error")
    return()
  }
  
  n_total <- nrow(fasta_df)
  chunks  <- split(fasta_df, ceiling(seq_len(n_total) / CD_CHUNK))
  n_chunk <- length(chunks)
  
  # 每个任务独立子目录（按上传文件名）
  task_id <- make_task_id(input$cd_search_fasta$name)
  out_dir <- file.path(CD_CACHE_DIR, task_id)
  dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
  message("[DEBUG] cd_search: task dir = ", out_dir)
  
  # PowerShell 可实时跟踪的纯文本进度文件
  progress_file <- file.path(out_dir, "progress.txt")
  log_progress <- function(msg) {
    cat(sprintf("[%s] %s\n", format(Sys.time(), "%Y-%m-%d %H:%M:%S"), msg),
        file = progress_file, append = TRUE)
  }
  log_progress(sprintf("==== START: %d 序列 / %d 批 (chunk=%d) ====",
                       n_total, n_chunk, CD_CHUNK))
  
  progress <- shiny::Progress$new()
  progress$set(message = sprintf("CD-Search: %d 序列 / %d 批", n_total, n_chunk), value = 0)
  on.exit(progress$close())
  
  collected <- vector("list", n_chunk)
  failed    <- integer(0)
  
  for (i in seq_len(n_chunk)) {
    progress$set(value = (i - 1) / n_chunk, detail = sprintf("Chunk %d/%d ...", i, n_chunk))
    
    cache_file <- file.path(out_dir, sprintf("chunk_%03d.rds", i))
    if (file.exists(cache_file)) {
      collected[[i]] <- readRDS(cache_file)
      message("[DEBUG] chunk ", i, " 命中缓存，跳过")
      log_progress(sprintf("chunk %d/%d  SKIP (已缓存)", i, n_chunk))
      next
    }
    
    ck <- chunks[[i]]
    tf <- tempfile(fileext = ".fasta")
    writeLines(paste0(">", ck$ID, "\n", ck$Sequence), tf)
    
    sub <- cd_submit(tf)
    if (file.exists(tf)) file.remove(tf)
    if (!sub$success) {
      message("[ERROR] chunk ", i, " 提交失败: ", sub$error)
      log_progress(sprintf("chunk %d/%d  FAIL (submit): %s", i, n_chunk, sub$error))
      failed <- c(failed, i); next
    }
    message("[DEBUG] chunk ", i, " cdsid = ", sub$cdsid)
    log_progress(sprintf("chunk %d/%d  submitted cdsid=%s", i, n_chunk, sub$cdsid))
    
    res <- cd_poll_fetch(sub$cdsid)
    if (!res$success) {
      message("[ERROR] chunk ", i, " 取结果失败: ", res$error)
      log_progress(sprintf("chunk %d/%d  FAIL (fetch): %s", i, n_chunk, res$error))
      failed <- c(failed, i); next
    }
    
    df <- cd_parse_tsv(res$content)
    if (is.null(df)) {
      message("[ERROR] chunk ", i, " 解析为空")
      log_progress(sprintf("chunk %d/%d  FAIL (parse)", i, n_chunk))
      failed <- c(failed, i); next
    }
    
    saveRDS(df, cache_file)
    collected[[i]] <- df
    log_progress(sprintf("chunk %d/%d  DONE  rows=%d", i, n_chunk, nrow(df)))
    Sys.sleep(10)
  }
  
  collected <- collected[!vapply(collected, is.null, logical(1))]
  if (length(collected) == 0) {
    log_progress("==== END: 全部失败 ====")
    showNotification("❌ 所有批次都失败了，请检查 proxy / captive portal。",
                     type = "error", duration = 15)
    return()
  }
  
  cd_all <- tryCatch(do.call(rbind, collected),
                     error = function(e) {
                       cols <- Reduce(intersect, lapply(collected, names))
                       do.call(rbind, lapply(collected, function(d) d[, cols, drop = FALSE]))
                     })
  
  rv$cd_auto_result <- cd_all
  progress$set(value = 1, detail = "Done!")
  
  # 跑完自动导出合并 TSV 到任务目录
  tsv_path <- file.path(out_dir, "cd_search_all.tsv")
  tryCatch({
    write.table(cd_all, tsv_path, sep = "\t", row.names = FALSE,
                quote = FALSE, fileEncoding = "UTF-8")
    message("[DEBUG] 已导出 TSV: ", tsv_path)
    log_progress(sprintf("==== END: %d 行，成功 %d/%d 批；TSV -> %s ====",
                         nrow(cd_all), length(collected), n_chunk, tsv_path))
  }, error = function(e) {
    message("[ERROR] TSV 导出失败: ", conditionMessage(e))
    log_progress(sprintf("==== END: %d 行，成功 %d/%d 批；TSV 导出失败: %s ====",
                         nrow(cd_all), length(collected), n_chunk, conditionMessage(e)))
  })
  
  msg <- sprintf("✅ CD-Search 完成：%d 行，成功 %d/%d 批；已导出 cd_search_all.tsv",
                 nrow(cd_all), length(collected), n_chunk)
  if (length(failed) > 0) {
    msg <- paste0(msg, "；失败批次: ", paste(failed, collapse = ", "),
                  "（缓存已保留，再次点击可续跑）")
  }
  showNotification(msg, type = if (length(failed) > 0) "warning" else "message", duration = 12)
})

# ---------- 手动上传（保持不变） ----------
cd_manual <- reactive({
  req(input$cd_manual_file)
  file_path <- input$cd_manual_file$datapath
  message("[DEBUG] cd_manual: reading file: ", file_path)
  tryCatch({
    line <- readLines(file_path, n = 1, warn = FALSE)
    sep  <- if (grepl("\t", line)) "\t" else ","
    df <- read.csv(file_path, sep = sep, stringsAsFactors = FALSE, check.names = FALSE, header = TRUE)
    if (ncol(df) < 2) {
      showNotification("The uploaded CD-Search result has too few columns.", type = "error")
      return(NULL)
    }
    message("[DEBUG] cd_manual: parsed, nrow = ", nrow(df), ", ncol = ", ncol(df))
    df
  }, error = function(e) {
    message("[ERROR] cd_manual: ", e$message)
    showNotification("Failed to read CD-Search result file.", type = "error")
    NULL
  })
})

cd_result <- reactive({
  if (!is.null(rv$cd_auto_result)) {
    message("[DEBUG] cd_result: using auto result")
    return(rv$cd_auto_result)
  }
  if (!is.null(cd_manual())) {
    message("[DEBUG] cd_result: using manual result")
    return(cd_manual())
  }
  NULL
})

output$cd_loaded <- reactive({ !is.null(cd_result()) })
outputOptions(output, "cd_loaded", suspendWhenHidden = FALSE)

output$cd_preview <- DT::renderDT({
  req(cd_result())
  DT::datatable(cd_result(), options = list(pageLength = 10, scrollX = TRUE), rownames = FALSE)
})

output$cd_search_status_ui <- renderUI({
  if (!is.null(cd_result())) {
    div(style = "color: #5cb85c;", icon("check-circle"),
        " CD-Search result loaded (", nrow(cd_result()), " rows)")
  } else {
    div(style = "color: #999;", "No CD-Search result yet.")
  }
})

output$cd_status_ui <- renderUI({
  if (is.null(cd_result())) {
    return(div(style = "margin-top: 10px; color: #999;", "No CD-Search annotation loaded."))
  }
  div(style = "margin-top: 10px; color: #5cb85c;",
      icon("check-circle"), " CD-Search annotation loaded: ", nrow(cd_result()), " entries")
})

add_cd_to_table <- function(protein_table, id_col_name = "Protein") {
  cds <- cd_result()
  if (is.null(cds) || nrow(cds) == 0) return(protein_table)
  if (!(id_col_name %in% colnames(protein_table))) return(protein_table)
  id_col_cd <- colnames(cds)[1]
  names(cds)[names(cds) == id_col_cd] <- id_col_name
  merge(protein_table, cds, by = id_col_name, all.x = TRUE, sort = FALSE)
}

output$download_cd_fasta <- downloadHandler(
  filename = function() paste0("cd_search_sequences_", format(Sys.time(), "%Y%m%d_%H%M%S"), ".fasta"),
  content = function(file) {
    fasta_df <- cd_fasta()
    if (is.null(fasta_df) || nrow(fasta_df) == 0) {
      writeLines(">No_sequences\n", file)
      return()
    }
    fasta_lines <- paste0(">", fasta_df$ID, "\n", fasta_df$Sequence)
    writeLines(fasta_lines, file)
  }
)

message("[DEBUG] cd_search.R loaded successfully (Final + auto TSV export)")

File: utils/cd_search.py
# utils/cd_search.py
# 完整版：自动分批提交 + 本地缓存 + 断点续跑 + 自动合并
# 支持从缓存加载已完成的结果
# 支持在新电脑上重新分批运行

import requests
import time
import tempfile
import os
import re
import glob
import json
from io import StringIO
import streamlit as st
import pandas as pd

DEBUG = True

def debug(msg: str):
    if DEBUG:
        timestamp = time.strftime("%H:%M:%S")
        print(f"[{timestamp}][CD-SEARCH] {msg}", flush=True)

# ==================== 配置 ====================
CD_ENDPOINT = "https://www.ncbi.nlm.nih.gov/Structure/bwrpsb/bwrpsb.cgi"
CD_CHUNK_SIZE = 950  # NCBI 限制每次不超过 1000 条
CD_MAX_WAIT = 900    # 每个批次最多等待 900 秒
CD_POLL_INTERVAL = 20  # 轮询间隔 20 秒

# ==================== 辅助函数 ====================

def get_cache_root():
    """获取 cd_cache 根目录的绝对路径"""
    current_dir = os.path.dirname(os.path.abspath(__file__))
    cache_root = os.path.join(current_dir, "..", "cd_cache")
    os.makedirs(cache_root, exist_ok=True)
    return cache_root


def make_task_id(fasta_text: str) -> str:
    """
    根据 FASTA 内容生成任务 ID。
    使用 SHA256 哈希，确保相同内容生成相同 ID。
    """
    import hashlib
    hash_val = hashlib.sha256(fasta_text.encode('utf-8')).hexdigest()[:16]
    return f"task_{hash_val}"


def parse_fasta_for_cd(fasta_text: str) -> tuple:
    """
    解析 FASTA 文本，返回 (ids, sequences) 列表。
    """
    debug("解析 FASTA 文本...")
    ids = []
    sequences = []
    current_id = None
    current_seq = []
    
    for line in fasta_text.splitlines():
        line = line.strip()
        if not line:
            continue
        if line.startswith(">"):
            if current_id is not None:
                ids.append(current_id)
                sequences.append("".join(current_seq))
            current_id = line[1:].split()[0]
            current_seq = []
        else:
            if current_id is not None:
                current_seq.append(line)
    
    if current_id is not None:
        ids.append(current_id)
        sequences.append("".join(current_seq))
    
    debug(f"解析完成：共 {len(ids)} 条序列")
    return ids, sequences


def split_into_chunks(ids: list, sequences: list, chunk_size: int = CD_CHUNK_SIZE) -> list:
    """
    将序列列表分割成多个批次。
    返回: [(chunk_ids, chunk_sequences, chunk_index), ...]
    """
    n_total = len(ids)
    n_chunks = (n_total + chunk_size - 1) // chunk_size
    chunks = []
    
    for i in range(n_chunks):
        start = i * chunk_size
        end = min(start + chunk_size, n_total)
        chunk_ids = ids[start:end]
        chunk_seqs = sequences[start:end]
        chunks.append((chunk_ids, chunk_seqs, i + 1))
    
    debug(f"分割完成：{n_total} 条序列 → {n_chunks} 个批次（每批最多 {chunk_size} 条）")
    return chunks


def write_fasta(ids: list, sequences: list, filepath: str):
    """写入 FASTA 文件"""
    with open(filepath, 'w', encoding='utf-8') as f:
        for seq_id, seq in zip(ids, sequences):
            f.write(f">{seq_id}\n{seq}\n")
    debug(f"写入 FASTA: {filepath} ({len(ids)} 条序列)")


def submit_cd_search(fasta_path: str) -> str:
    """
    提交 FASTA 文件到 NCBI CD-Search，返回 cdsid。
    失败返回 None。
    """
    debug(f"提交 CD-Search: {fasta_path}")
    
    try:
        with open(fasta_path, 'rb') as fp:
            params = {
                'db': 'cdd',
                'smode': 'auto',
                'tdata': 'hits',
                'dmode': 'rep',
                'cddefl': 'false',
                'qdefl': 'false',
                'clonly': 'false',
                'useid1': 'true',
            }
            files = {'queries': fp}
            resp = requests.post(CD_ENDPOINT, data=params, files=files, timeout=120)
        
        if resp.status_code != 200:
            debug(f"提交失败，HTTP {resp.status_code}")
            return None
        
        match = re.search(r'QM3-qcdsearch-[A-Za-z0-9-]+', resp.text)
        if not match:
            debug(f"未找到 CD-Search ID，响应前200字符: {resp.text[:200]}")
            return None
        
        cdsid = match.group(0)
        debug(f"提交成功，cdsid = {cdsid}")
        return cdsid
        
    except Exception as e:
        debug(f"提交异常: {e}")
        return None


def poll_and_fetch(cdsid: str, max_wait: int = CD_MAX_WAIT, interval: int = CD_POLL_INTERVAL) -> str:
    """
    轮询 CD-Search 结果，返回 TSV 文本内容。
    失败返回 None。
    """
    debug(f"开始轮询 {cdsid}，最多等待 {max_wait}s...")
    elapsed = 0
    
    while elapsed < max_wait:
        time.sleep(interval)
        elapsed += interval
        
        try:
            resp = requests.post(CD_ENDPOINT, data={'cdsid': cdsid, 'tdata': 'hits'}, timeout=120)
        except Exception as e:
            debug(f"轮询请求异常: {e}")
            continue
        
        status_lines = [line for line in resp.text.split('\n') if '#status' in line]
        if status_lines:
            try:
                status = int(re.search(r'(\d+)', status_lines[0]).group(1))
            except:
                status = -1
            
            if status == 0:
                debug(f"搜索完成（{elapsed}s）")
                return resp.text
            elif status == 3:
                debug(f"搜索进行中...（{elapsed}s）")
                continue
            else:
                debug(f"未知状态码: {status}（{elapsed}s）")
        
        # 检查是否有错误
        if 'error' in resp.text.lower() or 'too many' in resp.text.lower():
            debug(f"检测到错误响应: {resp.text[:300]}")
            return None
    
    debug(f"轮询超时（{max_wait}s）")
    return None


def parse_cd_response(content: str) -> pd.DataFrame:
    """
    解析 CD-Search 返回的 TSV 内容为 DataFrame。
    """
    lines = content.split('\n')
    data_lines = [l for l in lines if not l.startswith('#') and l.strip()]
    
    if not data_lines:
        debug("响应中无有效数据行")
        return None
    
    try:
        df = pd.read_csv(StringIO('\n'.join(data_lines)), sep='\t', dtype=str)
        # 清洗第一列（Query ID）
        id_col = df.columns[0]
        df[id_col] = df[id_col].str.replace(r'Q#\d+\s*-\s*>?', '', regex=True).str.strip()
        debug(f"解析成功：{len(df)} 条记录")
        return df
    except Exception as e:
        debug(f"解析 TSV 失败: {e}")
        return None


# ==================== 核心功能 ====================

def batch_cd_search_with_cache(fasta_text, progress_callback=None):
    """
    批量 CD-Search：自动分批 → 提交 → 缓存 → 合并。
    
    参数:
        fasta_text: FASTA 格式的序列文本
        progress_callback: 可选的回调函数，用于更新 Streamlit 进度
    
    返回:
        (DataFrame, task_dir): 合并后的结果和任务目录路径
    """
    debug("=" * 60)
    debug("开始批量 CD-Search（分批 + 缓存模式）")
    debug("=" * 60)
    
    # 解析 FASTA
    ids, sequences = parse_fasta_for_cd(fasta_text)
    n_total = len(ids)
    
    if n_total == 0:
        st.error("FASTA 中没有找到任何序列")
        return None, None
    
    # 创建任务目录
    task_id = make_task_id(fasta_text)
    cache_root = get_cache_root()
    task_dir = os.path.join(cache_root, task_id)
    os.makedirs(task_dir, exist_ok=True)
    debug(f"任务目录: {task_dir}")
    
    # 分割批次
    chunks = split_into_chunks(ids, sequences, CD_CHUNK_SIZE)
    n_chunks = len(chunks)
    
    # 进度文件
    progress_file = os.path.join(task_dir, "progress.txt")
    
    def log_progress(msg: str):
        """写入进度文件并打印到终端"""
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
        line = f"[{timestamp}] {msg}"
        with open(progress_file, 'a', encoding='utf-8') as f:
            f.write(line + '\n')
        debug(msg)
    
    log_progress(f"===== START: {n_total} 条序列 / {n_chunks} 个批次 =====")
    
    # 处理每个批次
    collected = []
    success_count = 0
    skip_count = 0
    fail_count = 0
    failed_chunks = []
    
    for chunk_idx, (chunk_ids, chunk_seqs, chunk_num) in enumerate(chunks):
        chunk_index = chunk_num  # 从1开始
        debug(f"\n--- 处理批次 {chunk_index}/{n_chunks} ({len(chunk_ids)} 条序列) ---")
        
        # 更新进度
        if progress_callback:
            progress_callback(chunk_index / n_chunks)
        
        # 检查缓存
        cache_file = os.path.join(task_dir, f"chunk_{chunk_index:04d}.tsv")
        
        if os.path.exists(cache_file):
            debug(f"批次 {chunk_index}: 发现缓存文件，直接加载")
            try:
                df = pd.read_csv(cache_file, sep='\t', dtype=str)
                collected.append(df)
                skip_count += 1
                log_progress(f"Chunk {chunk_index}/{n_chunks}: SKIP (cached, {len(df)} rows)")
                continue
            except Exception as e:
                debug(f"缓存文件损坏: {e}，将重新搜索")
                os.remove(cache_file)
        
        # 写入临时 FASTA
        temp_fasta = os.path.join(task_dir, f"chunk_{chunk_index:04d}.fasta")
        write_fasta(chunk_ids, chunk_seqs, temp_fasta)
        
        # 提交搜索
        cdsid = submit_cd_search(temp_fasta)
        
        # 删除临时 FASTA（节省空间）
        try:
            os.remove(temp_fasta)
        except:
            pass
        
        if cdsid is None:
            debug(f"批次 {chunk_index}: 提交失败")
            fail_count += 1
            failed_chunks.append(chunk_index)
            log_progress(f"Chunk {chunk_index}/{n_chunks}: FAIL (submit)")
            time.sleep(30)  # 失败后等待30秒再继续
            continue
        
        # 轮询结果
        content = poll_and_fetch(cdsid)
        
        if content is None:
            debug(f"批次 {chunk_index}: 获取结果失败")
            fail_count += 1
            failed_chunks.append(chunk_index)
            log_progress(f"Chunk {chunk_index}/{n_chunks}: FAIL (fetch)")
            time.sleep(30)
            continue
        
        # 解析结果
        df = parse_cd_response(content)
        
        if df is None or df.empty:
            debug(f"批次 {chunk_index}: 解析结果为空")
            fail_count += 1
            failed_chunks.append(chunk_index)
            log_progress(f"Chunk {chunk_index}/{n_chunks}: FAIL (parse)")
            time.sleep(30)
            continue
        
        # 保存缓存
        df.to_csv(cache_file, sep='\t', index=False)
        collected.append(df)
        success_count += 1
        log_progress(f"Chunk {chunk_index}/{n_chunks}: DONE ({len(df)} rows)")
        debug(f"缓存已保存: {cache_file}")
        
        # 批次之间休息 10 秒，避免请求过快
        time.sleep(10)
    
    # 最终汇总
    debug(f"\n===== 汇总 =====")
    debug(f"成功: {success_count}/{n_chunks}")
    debug(f"跳过(缓存): {skip_count}/{n_chunks}")
    debug(f"失败: {fail_count}/{n_chunks}")
    
    if not collected:
        log_progress("===== END: 所有批次均失败 =====")
        return None, task_dir
    
    # 合并所有结果
    debug("合并所有批次结果...")
    combined = pd.concat(collected, ignore_index=True)
    
    # 保存合并结果
    combined_file = os.path.join(task_dir, "cd_search_all.tsv")
    combined.to_csv(combined_file, sep='\t', index=False)
    debug(f"合并结果已保存: {combined_file}")
    debug(f"合并后共 {len(combined)} 条记录，{combined[combined.columns[0]].nunique()} 个唯一蛋白")
    
    log_progress(f"===== END: {len(combined)} rows, {success_count} success, {skip_count} skip, {fail_count} fail =====")
    
    if progress_callback:
        progress_callback(1.0)
    
    return combined, task_dir


def batch_cd_search_simple(fasta_text):
    """
    简化版：提交单个 FASTA 到 NCBI CD-Search（用于少量序列）。
    返回 DataFrame 或 None。
    """
    debug("开始自动 CD-Search（单次提交）...")
    
    if not fasta_text.strip():
        st.warning("FASTA 内容为空")
        return None
    
    with tempfile.NamedTemporaryFile(mode='w', suffix='.fasta', delete=False, encoding='utf-8') as f:
        f.write(fasta_text)
        temp_path = f.name
    
    debug(f"临时 FASTA: {temp_path}")
    
    try:
        cdsid = submit_cd_search(temp_path)
        if cdsid is None:
            return None
        
        content = poll_and_fetch(cdsid)
        if content is None:
            return None
        
        df = parse_cd_response(content)
        return df
        
    except Exception as e:
        debug(f"异常: {e}")
        return None
    finally:
        try:
            if os.path.exists(temp_path):
                os.unlink(temp_path)
        except:
            pass


# ==================== 手动文件解析 ====================

def parse_cd_tsv(content):
    """
    解析手动上传的 CD-Search TSV 文本。
    """
    debug("开始解析手动上传的 CD-Search TSV 文件")
    lines = content.split('\n')
    data_lines = [l for l in lines if not l.startswith('#') and l.strip()]
    if not data_lines:
        st.warning("文件中未找到有效数据行")
        return None

    try:
        sep = '\t' if '\t' in data_lines[0] else ','
        df = pd.read_csv(StringIO('\n'.join(data_lines)), sep=sep, dtype=str)
        debug(f"解析到 {df.shape[0]} 行, {df.shape[1]} 列")
        id_col = df.columns[0]
        df[id_col] = df[id_col].str.replace(r'Q#\d+\s*-\s*>?', '', regex=True).str.strip()
        debug(f"清洗后共 {len(df)} 条记录")
        return df
    except Exception as e:
        st.error(f"解析 TSV 时出错: {e}")
        return None


# ==================== 缓存加载 ====================

def load_cd_cache(task_name: str = None):
    """
    从本地 cd_cache/ 目录加载已完成的 CD-Search 结果。
    """
    cache_root = get_cache_root()
    debug(f"扫描缓存目录: {cache_root}")
    
    if not os.path.exists(cache_root):
        debug("缓存目录不存在")
        return None, None
    
    subdirs = [d for d in os.listdir(cache_root) if os.path.isdir(os.path.join(cache_root, d))]
    debug(f"发现 {len(subdirs)} 个子目录: {subdirs}")
    
    if not subdirs:
        return None, None
    
    if task_name is not None:
        if task_name not in subdirs:
            debug(f"任务 '{task_name}' 不在缓存中")
            return None, None
        subdirs = [task_name]
    
    all_chunks = []
    loaded = 0
    failed = 0
    
    for subdir in subdirs:
        subdir_path = os.path.join(cache_root, subdir)
        
        # 优先使用合并文件
        combined_file = os.path.join(subdir_path, "cd_search_all.tsv")
        if os.path.exists(combined_file):
            try:
                df = pd.read_csv(combined_file, sep='\t', dtype=str)
                debug(f"加载合并文件: {combined_file} ({len(df)} rows)")
                all_chunks.append(df)
                loaded += 1
                continue
            except Exception as e:
                debug(f"读取合并文件失败: {e}")
        
        # 回退：加载各个分块文件
        chunk_files = sorted(glob.glob(os.path.join(subdir_path, "chunk_*.tsv")))
        debug(f"目录 {subdir}: 找到 {len(chunk_files)} 个分块文件")
        
        for cf in chunk_files:
            try:
                df = pd.read_csv(cf, sep='\t', dtype=str)
                all_chunks.append(df)
                loaded += 1
            except Exception as e:
                debug(f"读取 {cf} 失败: {e}")
                failed += 1
    
    debug(f"加载完成: 成功 {loaded}, 失败 {failed}")
    
    if not all_chunks:
        return None, None
    
    combined = pd.concat(all_chunks, ignore_index=True)
    debug(f"合并后共 {len(combined)} 条记录")
    
    task_used = task_name if task_name else subdirs[0]
    return combined, task_used


def get_cached_tasks():
    """获取所有可用的缓存任务名称"""
    cache_root = get_cache_root()
    if not os.path.exists(cache_root):
        return []
    tasks = [d for d in os.listdir(cache_root) if os.path.isdir(os.path.join(cache_root, d))]
    debug(f"可用缓存任务: {tasks}")
    return tasks
